File: code/19-Remove Nth Node From End of List/19-Remove Nth Node From End of List.py
"""
LeetCode 19 - Remove Nth Node From End of List

解题步骤（自动整理）
1. 用快慢指针/虚拟头节点等技巧处理链表边界
2. 按题意进行遍历、反转、合并或断链等操作
3. 返回处理后的链表头节点或计算结果
"""
# Given a linked list, remove the n-th node from the end of list and return its head.
#
# Example:
#
# Given linked list: 1->2->3->4->5, and n = 2.
#
# After removing the second node from the end, the linked list becomes 1->2->3->5.
# Note:
#
# Given n will always be valid.
#
# Follow up:
#
# Could you do this in one pass?


# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        pre = ListNode(-1)
        pre.next = head
        curr, end = pre, pre
        for i in range(n):
            end = end.next

        while end and end.next:
            curr = curr.next
            end = end.next
        if n != 1:
            curr.next = curr.next.next
        else:
            curr.next = None
        return pre.next

File: code/19-Remove Nth Node From End of List/19-Remove Nth Node From End of List.c
/*
LeetCode 19 - Remove Nth Node From End of List

解题步骤（自动整理）
1. 用快慢指针/虚拟头节点等技巧处理链表边界
2. 按题意进行遍历、反转、合并或断链等操作
3. 返回处理后的链表头节点或计算结果
*/
/*
Remove Nth Node From End of List

Given a linked list, remove the nth node from the end of list and return its head.

For example,

   Given linked list: 1->2->3->4->5, and n = 2.

   After removing the second node from the end, the linked list becomes 1->2->3->5.
Note:
Given n will always be valid.
Try to do this in one pass.
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */


struct ListNode* removeNthFromEnd(struct ListNode* head, int n){
    struct ListNode *pre = head, *end = head;
    for (int i = 0; i < n; i ++) {
        end = end->next;
        if (!end) {
            return head->next;
        }
    }
    while (end->next) {
        pre = pre->next;
        end = end->next;
    }
    pre->next = pre->next->next;
    return head;
}
